File: backend/handlers/groups.go
package handlers

import (
	"net/http"
)

// GroupHandler handles all HTTP endpoints related to groups and community features.
// Groups are a visibility layer between friends (private social) and public (world map).
type GroupHandler struct {
	db interface{} // will be *pgxpool.Pool once implemented
}

// NewGroupHandler creates a new GroupHandler ready to register with the router.
func NewGroupHandler(db interface{}) *GroupHandler {
	return &GroupHandler{db: db}
}

// CreateGroup handles creating a new group.
// The authenticated user becomes the group's first admin member automatically.
// The request body should include name, description, and whether the group is private.
func (h *GroupHandler) CreateGroup(w http.ResponseWriter, r *http.Request) {
}

// GetGroup returns the full details of a single group by its ID, including
// the current member count. Requires the requester to be a member of the group
// if the group is private.
func (h *GroupHandler) GetGroup(w http.ResponseWriter, r *http.Request) {
}

// UpdateGroup lets an admin change the group's name, description, or privacy setting.
// Non-admin members get a 403. Changing from private to public automatically
// expires any outstanding invite code.
func (h *GroupHandler) UpdateGroup(w http.ResponseWriter, r *http.Request) {
}

// DeleteGroup permanently removes the group and all its memberships.
// Only the admin can do this. All catches shared exclusively to this group
// remain in the database but become effectively invisible to other group members.
func (h *GroupHandler) DeleteGroup(w http.ResponseWriter, r *http.Request) {
}

// JoinGroup lets a user join a public group immediately without needing approval.
// A new GroupMember row is inserted with the "member" role.
// Returns 409 if the user is already a member, 403 if the group is private
// (private groups require an invite code or a join request instead).
func (h *GroupHandler) JoinGroup(w http.ResponseWriter, r *http.Request) {
}

// RequestToJoinGroup creates a pending JoinRequest for a private group.
// The request waits until an admin or moderator accepts or declines it.
// Notifying the group's admins is handled separately by the notification job.
func (h *GroupHandler) RequestToJoinGroup(w http.ResponseWriter, r *http.Request) {
}

// AcceptJoinRequest approves a pending join request, converting it to an active
// GroupMember row with the "member" role. Only admins and moderators can do this.
// The status on the JoinRequest is updated to "accepted" for audit trail purposes.
func (h *GroupHandler) AcceptJoinRequest(w http.ResponseWriter, r *http.Request) {
}

// DeclineJoinRequest rejects a pending join request. Only admins and moderators
// can do this. The status is updated to "declined" so the requester can see
// their request was reviewed rather than just ignored.
func (h *GroupHandler) DeclineJoinRequest(w http.ResponseWriter, r *http.Request) {
}

// LeaveGroup removes the authenticated user from a group they currently belong to.
// If the leaving member is the sole admin, the request is rejected — the group
// must be deleted or another member must be promoted to admin first.
func (h *GroupHandler) LeaveGroup(w http.ResponseWriter, r *http.Request) {
}

// GenerateInviteCode creates a new short random invite code for a private group
// and saves it on the Group record, replacing any previous code. Only admins can
// call this. The code lets anyone with the link bypass the join-request flow.
func (h *GroupHandler) GenerateInviteCode(w http.ResponseWriter, r *http.Request) {
}

// JoinByInviteCode lets a user join a private group by supplying a valid invite code.
// The server looks up the group by the code, verifies it matches, and adds the user
// as a regular member. Returns 404 if no group has that code.
func (h *GroupHandler) JoinByInviteCode(w http.ResponseWriter, r *http.Request) {
}

// RemoveMember lets an admin or moderator remove another user from the group.
// Admins can remove anyone. Moderators can only remove regular members, not other
// moderators or admins. The removed user's catches stay in the database.
func (h *GroupHandler) RemoveMember(w http.ResponseWriter, r *http.Request) {
}

// PromoteMember elevates a regular member to the moderator role.
// Only admins can promote. Moderators cannot promote other members.
func (h *GroupHandler) PromoteMember(w http.ResponseWriter, r *http.Request) {
}

// DemoteMember drops a moderator back down to the regular member role.
// Only admins can demote. Returns 400 if the target is already a regular member.
func (h *GroupHandler) DemoteMember(w http.ResponseWriter, r *http.Request) {
}

// ListGroupMembers returns a paginated list of all active members in the group
// along with each member's username and role. Available to any group member.
// Pagination uses cursor-based approach via a "page" query parameter.
func (h *GroupHandler) ListGroupMembers(w http.ResponseWriter, r *http.Request) {
}

// ListUserGroups returns all groups the authenticated user currently belongs to,
// including their role in each group. Used to populate the groups tab and to
// build the group picker when logging a new catch.
func (h *GroupHandler) ListUserGroups(w http.ResponseWriter, r *http.Request) {
}

// GetGroupFeed returns a paginated list of catches that have been shared to this
// specific group, ordered newest-first. Only members of the group can see this feed.
// Catches shared to multiple groups each appear once per group feed.
func (h *GroupHandler) GetGroupFeed(w http.ResponseWriter, r *http.Request) {
}

// GetGroupStats returns aggregate statistics for the group: total number of catches,
// the top species caught, the most active members by catch count, and a time-series
// of catch rate over the past 30 days. Only available to group members.
func (h *GroupHandler) GetGroupStats(w http.ResponseWriter, r *http.Request) {
}

// SearchPublicGroups lets any authenticated user discover public groups by name
// or by a location string (e.g. a lake or region name stored in the description).
// Returns a paginated list of matching groups. Private groups are excluded entirely.
func (h *GroupHandler) SearchPublicGroups(w http.ResponseWriter, r *http.Request) {
}

File: frontend-shared/api/groups.ts
/**
 * Groups API functions.
 * Manages group creation, membership, moderation, and group-scoped catch data.
 */

import type { ApiClient } from './client';
import type {
  Group,
  GroupMember,
  JoinRequest,
  CreateGroupRequest,
  UpdateGroupRequest,
  GroupFeedResponse,
  GroupStats,
} from '../types/group';

// Creates a new group. The authenticated user automatically becomes the admin.
// Returns the newly created group record.
export async function createGroup(
  client: ApiClient,
  body: CreateGroupRequest,
): Promise<Group> {
  return client.post<Group>('/groups', body);
}

// Returns the details of a single group by its ID.
// Private groups return 403 if the requester is not a member.
export async function getGroup(
  client: ApiClient,
  groupId: number,
): Promise<Group> {
  return client.get<Group>(`/groups/${groupId}`);
}

// Updates a group's name, description, or privacy setting.
// Only admins of the group can call this. Returns the updated group.
export async function updateGroup(
  client: ApiClient,
  groupId: number,
  body: UpdateGroupRequest,
): Promise<Group> {
  return client.put<Group>(`/groups/${groupId}`, body);
}

// Permanently deletes a group and all its memberships.
// Only the group admin can call this.
export async function deleteGroup(
  client: ApiClient,
  groupId: number,
): Promise<void> {
  return client.delete<void>(`/groups/${groupId}`);
}

// Joins a public group immediately. Returns the new GroupMember record.
// Returns 403 for private groups (use requestToJoinGroup or joinByInviteCode instead).
export async function joinGroup(
  client: ApiClient,
  groupId: number,
): Promise<GroupMember> {
  return client.post<GroupMember>(`/groups/${groupId}/join`);
}

// Submits a join request for a private group. Returns the pending JoinRequest.
// An admin or moderator must approve the request before the user becomes a member.
export async function requestToJoinGroup(
  client: ApiClient,
  groupId: number,
): Promise<JoinRequest> {
  return client.post<JoinRequest>(`/groups/${groupId}/request`);
}

// Approves a pending join request. The user becomes a regular member.
// Only admins and moderators can call this.
export async function acceptJoinRequest(
  client: ApiClient,
  groupId: number,
  requestId: number,
): Promise<GroupMember> {
  return client.post<GroupMember>(`/groups/${groupId}/requests/${requestId}/accept`);
}

// Declines a pending join request. The status is updated to "declined".
// Only admins and moderators can call this.
export async function declineJoinRequest(
  client: ApiClient,
  groupId: number,
  requestId: number,
): Promise<void> {
  return client.post<void>(`/groups/${groupId}/requests/${requestId}/decline`);
}

// Removes the authenticated user from the group.
// Returns 400 if the user is the sole admin (must transfer ownership first).
export async function leaveGroup(
  client: ApiClient,
  groupId: number,
): Promise<void> {
  return client.post<void>(`/groups/${groupId}/leave`);
}

// Generates a new invite code for a private group, replacing any existing one.
// Returns the updated group record with the new invite_code field.
// Only the group admin can call this.
export async function generateInviteCode(
  client: ApiClient,
  groupId: number,
): Promise<Group> {
  return client.post<Group>(`/groups/${groupId}/invite-code`);
}

// Joins a private group using a valid invite code, bypassing the join-request flow.
// Returns the new GroupMember record. Returns 404 if the code is invalid.
export async function joinByInviteCode(
  client: ApiClient,
  inviteCode: string,
): Promise<GroupMember> {
  return client.post<GroupMember>('/groups/join-by-invite', { invite_code: inviteCode });
}

// Removes a specific member from the group.
// Admins can remove anyone; moderators can only remove regular members.
export async function removeMember(
  client: ApiClient,
  groupId: number,
  userId: number,
): Promise<void> {
  return client.delete<void>(`/groups/${groupId}/members/${userId}`);
}

// Promotes a regular member to the moderator role.
// Only the group admin can call this.
export async function promoteMember(
  client: ApiClient,
  groupId: number,
  userId: number,
): Promise<GroupMember> {
  return client.post<GroupMember>(`/groups/${groupId}/members/${userId}/promote`);
}

// Demotes a moderator back to the regular member role.
// Only the group admin can call this.
export async function demoteMember(
  client: ApiClient,
  groupId: number,
  userId: number,
): Promise<GroupMember> {
  return client.post<GroupMember>(`/groups/${groupId}/members/${userId}/demote`);
}

// Returns a paginated list of the group's active members with their roles.
// Available to any member of the group.
export async function listGroupMembers(
  client: ApiClient,
  groupId: number,
  page = 1,
): Promise<{ members: GroupMember[]; total: number }> {
  return client.get<{ members: GroupMember[]; total: number }>(
    `/groups/${groupId}/members?page=${page}`,
  );
}

// Returns all groups the authenticated user currently belongs to, with their role in each.
// Used to populate the groups tab and the group picker when logging a catch.
export async function listUserGroups(client: ApiClient): Promise<GroupMember[]> {
  return client.get<GroupMember[]>('/groups/mine');
}

// Returns a paginated feed of catches shared to a specific group, newest first.
// Only members of the group can call this.
export async function getGroupFeed(
  client: ApiClient,
  groupId: number,
  page = 1,
): Promise<GroupFeedResponse> {
  return client.get<GroupFeedResponse>(`/groups/${groupId}/feed?page=${page}`);
}

// Returns aggregate statistics for the group: total catches, top species,
// most active members, and a catch-rate time series over the past 30 days.
// Only available to group members.
export async function getGroupStats(
  client: ApiClient,
  groupId: number,
): Promise<GroupStats> {
  return client.get<GroupStats>(`/groups/${groupId}/stats`);
}

// Searches for public groups by name or location keyword.
// Returns a paginated list of matching groups. Private groups are excluded.
export async function searchPublicGroups(
  client: ApiClient,
  query: string,
  page = 1,
): Promise<{ groups: Group[]; total: number }> {
  return client.get<{ groups: Group[]; total: number }>(
    `/groups/search?q=${encodeURIComponent(query)}&page=${page}`,
  );
}
